File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/finder/MyCanalFinder.java
package lan.dk.podcastserver.manager.worker.finder;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.jayway.jsonpath.DocumentContext;
import io.vavr.collection.List;
import io.vavr.control.Option;
import io.vavr.control.Try;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.service.HtmlService;
import lan.dk.podcastserver.service.ImageService;
import lan.dk.podcastserver.service.JsonService;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.hibernate.validator.constraints.NotEmpty;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.springframework.stereotype.Service;

import static io.vavr.API.*;

/**
 * Created by kevin on 16/03/2016 for Podcast Server
 */
@Slf4j
@Service("MyCanalFinder")
@RequiredArgsConstructor
public class MyCanalFinder implements Finder {

    private static final String DOMAIN = "https://www.mycanal.fr";
    private final HtmlService htmlService;
    private final ImageService imageService;
    private final JsonService jsonService;

    @Override
    public Podcast find(String url) {
        Option<DocumentContext> json = htmlService
                .get(url)
                .map(Document::body)
                .flatMap(es -> List.ofAll(es.select("script")).find(e -> e.html().contains("app_config")))
                .map(Element::html)
                .flatMap(MyCanalFinder::extractJsonConfig)
                .map(jsonService::parse);

        return json.toTry().map(JsonService.to("landing.cover", MyCanalLandingItem.class)).map(this::asPodcast)
                .orElse(() -> json.toTry().map(JsonService.to("page", MyCanalPageItem.class)).map(this::asPodcast))
                .getOrElse(Podcast.DEFAULT_PODCAST);
    }

     private Podcast asPodcast(MyCanalLandingItem item) {
        return Podcast.builder()
                .title(item.getOnClick().getDisplayName())
                .url(DOMAIN + item.getOnClick().getPath())
                .cover(imageService.getCoverFromURL(item.getImage()))
                .type("MyCanal")
                .build();
    }

    private Podcast asPodcast(MyCanalPageItem item) {
        return Podcast.builder()
                .title(item.getDisplayName())
                .url(DOMAIN + item.getPathname())
                .cover(Cover.DEFAULT_COVER)
                .type("MyCanal")
                .build();
    }

    public static Option<String> extractJsonConfig(String text) {
        String startToken = "__data=";
        String endToken = "};";

        if (!text.contains(startToken) || !text.contains(endToken)) {
            log.error("Structure of MyCanal page changed");
            return None();
        }

        int begin = text.indexOf(startToken);
        int end = text.indexOf(endToken, begin);
        return Option(text.substring(begin + startToken.length(), end + 1));
    }

    @Override
    public Integer compatibility(@NotEmpty String url) {
        return _compatibility(url);
    }

    public static int _compatibility(@NotEmpty String url) {
        return StringUtils.contains(url, "www.mycanal.fr") ? 1 : Integer.MAX_VALUE;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalLandingItem {
        @Getter @Setter String image;
        @Getter @Setter MyCanalOnClick onClick;

        @JsonIgnoreProperties(ignoreUnknown = true)
        private static class MyCanalOnClick {
            @Getter @Setter String displayName;
            @Getter @Setter String path;
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class MyCanalPageItem {
        @Getter @Setter String displayName;
        @Getter @Setter String pathname;
    }



}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/mycanal/MyCanalFinder.kt
package com.github.davinkevin.podcastserver.manager.worker.mycanal

import arrow.core.Try
import arrow.core.getOrElse
import arrow.core.orElse
import arrow.syntax.collections.firstOption
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.utils.k
import lan.dk.podcastserver.entity.Cover
import lan.dk.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.manager.worker.Finder
import lan.dk.podcastserver.service.JsonService
import org.springframework.stereotype.Service

/**
 * Created by kevin on 16/03/2016 for Podcast Server
 */
@Service
class MyCanalFinder(val htmlService: HtmlService, val imageService: ImageService, val jsonService: JsonService) : Finder {

    override fun find(url: String): Podcast {
        val json = Try {
            htmlService.get(url).k()
                    .map { it.body() }
                    .flatMap { es -> es.select("script").firstOption { it.html().contains("app_config") } }
                    .map { it.html() }
                    .flatMap { extractJsonConfig(it).k() }
                    .map { jsonService.parse(it) }
                    .getOrElse { throw RuntimeException("Extraction of app_config ends up in error") }
        }

        return json
                .flatMap { Try { JsonService.to("landing.cover", MyCanalItem::class.java).apply(it) } }
                .map { toPodcast(it) }
                .orElse { json
                        .flatMap { Try { JsonService.to ("page", MyCanalPageItem::class.java).apply(it) } }
                        .map { toPodcast(it) }
                }
                .getOrElse { Podcast.DEFAULT_PODCAST }
    }

    private fun toPodcast(item: MyCanalItem) =
            Podcast().apply {
                title = item.onClick.displayName
                url = DOMAIN + item.onClick.path
                cover = imageService.getCoverFromURL(item.image)
                type = "MyCanal"
            }

    private fun toPodcast(item: MyCanalPageItem) =
            Podcast().apply {
                title = item.displayName
                url = DOMAIN + item.pathname
                cover = Cover.DEFAULT_COVER
                type = "MyCanal"
            }

    override fun compatibility(url: String?) = myCanalCompatibility(url)

    companion object {
        private const val DOMAIN = "https://www.mycanal.fr"
    }

}
